automata/parse/lexer.py

```diff
@@ -0,0 +1,160 @@
+import re
+import abc
+from dataclasses import dataclass
+from typing import Callable, List, Optional, Set, Tuple, TypeVar, Generic
+
+ResultT = TypeVar('ResultT')
+@dataclass
+class LexerError(Exception):
+    """An exception raised for issues in lexing"""
+    message: str
+    position: int
+
+
+class Token(Generic[ResultT], metaclass=abc.ABCMeta):
+    """Base class for tokens."""
+
+    __slots__ = ['text']
+
+    text: str
+
+    def __init__(self, text: str) -> None:
+        self.text = text
+
+    def get_precedence(self) -> int:
+        raise NotImplementedError
+
+    def __repr__(self) -> str:
+        return f"<{self.__class__.__name__}: {self.text}>"
+
+TokenFactoryT = Callable[[str], Token[ResultT]]
+
+class TokenRegistry(Generic[ResultT]):
+    """Holds a bunch of token rules.
+
+    Attributes:
+        _tokens ((Token, re) list): the registered tokens.
+    """
+
+    __slots__ = ['_tokens']
+
+    def __init__(self) -> None:
+        self._tokens = []
+
+    def register(self, token_factory_fn: TokenFactoryT, token_regex: str) -> None:
+        """Register a token.
+
+        Args:
+            token (Token): the token class to register
+            regexp (str): the regexp for that token
+        """
+        self._tokens.append((token_factory_fn, re.compile(token_regex)))
+
+    def matching_tokens(self, text: str, start: int):
+        """Retrieve all token definitions matching the beginning of a text.
+
+        Args:
+            text (str): the text to test
+            start (int): the position where matches should be searched in the
+                string (see re.match(rx, txt, pos))
+
+        Yields:
+            (token_class, re.Match): all token class whose regexp matches the
+                text, and the related re.Match object.
+        """
+        res = []
+
+        for token_factory_fn, regexp in self._tokens:
+            match = regexp.match(text, pos=start)
+            if match:
+                res.append((token_factory_fn, match))
```

Comment:
This function originally yielded, but I switched to a list. On a second look I'll probably go ahead and switch it back (since I think you're right that yielding is faster / more flexible).

Reply:
@eliotwrobson Sounds good to me. Please proceed.